### ISSUE TITLE
Support for grid metrics

### ISSUE THREAD

**Issue by jbusecke** [xgcm/xgcm#81]
Ok now that #74 is merged 🎆 🍾 , I would like to talk about the next step: Recalculating distances from given coordinates.

This is much easier to implement, I believe. However, I think it is worth talking about the actual calculation in detail.
The case worth discussing is the calculation of physical [meters] from positions on the globe, since distances for e.g. depth (already in meters) are now easy to implement using the xgcm.Grid/Axis diff function.

I put together a simple function to calculate the distance in meters from dlon/dlat (longitude/latitude differentials in deg) and the actual longitude/latitude.

```python
    def dll_dist(dlon, dlat, lon, lat):
        """Converts lat/lon differentials into distances
   
        PARAMETERS
        ----------
        dlon : xarray.DataArray longitude differentials
        dlat : xarray.DataArray latitude differentials
        lon  : xarray.DataArray longitude values
        lat  : xarray.DataArray latitude values
   
        RETURNS
        -------
        dx  : xarray.DataArray distance inferred from dlon
        dy  : xarray.DataArray distance inferred from dlat
        """
   
        # First attempt with super cheap approach...
        #     111km for each deg lat and then scale that by cos(lat) for lon
        dll_factor = 111000.0
        dx = dlon * np.cos(np.deg2rad(lat.data)) * dll_factor
        dy = dlat * dll_factor
   
        return dx, dy
```
As @rabernat mentioned in #74 this assumes a projection already (Mercator I guess?). Is there a sleek way to do this with existing packages like cartopy?

**Comment by rabernat** [comment]
Before we get to this issue, I think it is important to think about how xgcm will interact with these "[metrics](https://en.wikipedia.org/wiki/Metric_(mathematics))" (seems like a good term) in more generality. Many datasets already come with them defined. How do the metrics related to `Grid` an `Axis` objects?

The metrics define the distance between axis. Inspired by MITgcm, I imagine metrics `dxc` and `dxg` associated with an `Axis` in the following way:

```
|------o-------|------o-------|------o-------|------o-------|
<-----dxg------>
       <-----dxc------>
```

However, one could make this higher order by doing
```
|------o-------|------o-------|------o-------|------o-------|
 <-dxl-><-dxr->
```
(I believe this is the approach used by MOM.)

By associating each `Axis` with the appropriate metrics, we can turn our difference operation into a derivative, and our sum / cumsum into an integration. (The metrics also open the door to more accurate interpolation methods, such as those used in MOM.)

I believe that the metrics are not dependent on a projection. They represent the great-circle distance between any two points on the Earth. (Your earlier code defined `x` and `y` in addition to `dx` and `dy`; that is a crucial difference.) The `dx` and `dy` you defined above are approximations to great-circle distance for closely spaced lon and lat points.

For cartesian geometry, the area and volume metrics are just products of the axis metrics. For spherical geometry, they are not. One could imaging storing metrics in `Grid` object as a dictionary keyed by sets of dimensions, for example
```python
{ {'XC', 'YC'}: rA, {'XG', 'YG}: rAz }
```
(It can be useful to see how a GCM puts together these metrics: https://github.com/altMITgcm/MITgcm/blob/master/model/src/ini_spherical_polar_grid.F)

Only once we have a good design for the metrics themselves will we be ready to tackle autogeneration of such metrics. So this issue might have to wait for that to be resolved first.

**Comment by jbusecke** [comment]
I think that we should go the higher-order way, at least internally. It should be pretty simple to convert from the 'left/right' distances `dxr``dxl` to `dxg` in the above example. 
From what I understand that is also what is done in the MOM5 code. 
Converting the other way might need some more assumptions. 

Expanding what I suggested in #99, I think internally we need to store something like `{{'XC', 'YC'}: {'area':'rA', 'dx':'dxc', 'dxr':[], 'dxl':[]},...}`
This means for each combination of dimensions in the dataset we need a dictionary with the following values:
- xdistance (e.g. `dxg` above)
-  2 'quadrant' xdistances (e.g. 'dxl' and 'dxr' above)
- ydistance
-  2 'quadrant' ydistances
- zdistance
-  2 'quadrant' zdistances

Then we would need to store areas and 'quadrant' areas and the same with volumes. 
Potentially we also need the information about the partial cells or some other 'mask' parameter?

This is a lot of stuff but I think if we succeed, that would really be the beauty of having it in xgcm: Define all metrics once and then just science along without worrying about it.

If any of these values are missing we eventually will have the logic to trigger a recalculation at the time of grid object initialization or as needed by e.g. the calculus functions.
